From af6d0a2bfaa4b9b27100f955cad1d27c0a509be8 Mon Sep 17 00:00:00 2001
From: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
Date: Thu, 16 Oct 2025 09:15:24 +0200
Subject: [PATCH 172/239] VectorImage: Fix offset transformation of use nodes

Instead of just using transform property itself for this, the
<use> node in SVG has an additional offset property which is documented
to be a "final offset transformation".

We would previously apply this as the x and y coordinates of the item
in the generated Qt Quick code, which works when there is no other
transform on the item (or the only transform is another translation).

But since the x and y is applied before the other transforms,
if we scale, rotate, etc. the coordinate system, then we would get
the wrong position.

Change-Id: I72b05439c06619231cfc5d4ba5a639e06d7aa184
Reviewed-by: Hatem ElKharashy <hatem.elkharashy@qt.io>
Reviewed-by: Eirik Aavitsland <eirik.aavitsland@qt.io>
(cherry picked from commit 22ac54787fb07b19e55a0a0f5d1af3c727817653)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/quickvectorimage/generator/qquicknodeinfo_p.h     |  1 -
 src/quickvectorimage/generator/qquickqmlgenerator.cpp |  2 --
 src/quickvectorimage/generator/qsvgvisitorimpl.cpp    | 11 ++++++++++-
 3 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/src/quickvectorimage/generator/qquicknodeinfo_p.h b/src/quickvectorimage/generator/qquicknodeinfo_p.h
index 468f20117e..d6de8176e9 100644
--- a/src/quickvectorimage/generator/qquicknodeinfo_p.h
+++ b/src/quickvectorimage/generator/qquicknodeinfo_p.h
@@ -121,7 +121,6 @@ enum class StructureNodeStage
 
 struct UseNodeInfo : NodeInfo
 {
-    QPointF startPos;
     StructureNodeStage stage;
 };
 
diff --git a/src/quickvectorimage/generator/qquickqmlgenerator.cpp b/src/quickvectorimage/generator/qquickqmlgenerator.cpp
index 411d1a4f5d..df41e4f132 100644
--- a/src/quickvectorimage/generator/qquickqmlgenerator.cpp
+++ b/src/quickvectorimage/generator/qquickqmlgenerator.cpp
@@ -736,8 +736,6 @@ void QQuickQmlGenerator::generateUseNode(const UseNodeInfo &info)
         stream() << "Item {";
         m_indentLevel++;
         generateNodeBase(info);
-        stream() << "x: " << info.startPos.x();
-        stream() << "y: " << info.startPos.y();
     } else {
         m_indentLevel--;
         stream() << "}";
diff --git a/src/quickvectorimage/generator/qsvgvisitorimpl.cpp b/src/quickvectorimage/generator/qsvgvisitorimpl.cpp
index 012a6209a4..87ce1e80ff 100644
--- a/src/quickvectorimage/generator/qsvgvisitorimpl.cpp
+++ b/src/quickvectorimage/generator/qsvgvisitorimpl.cpp
@@ -996,7 +996,16 @@ void QSvgVisitorImpl::visitUseNode(const QSvgUse *node)
     fillAnimationInfo(node, info);
 
     info.stage = StructureNodeStage::Start;
-    info.startPos = node->start();
+
+    QPointF startPos = node->start();
+    if (!startPos.isNull()) {
+        QTransform xform;
+        if (!info.isDefaultTransform)
+            xform = info.transform.defaultValue().value<QTransform>();
+        xform.translate(startPos.x(), startPos.y());
+        info.transform.setDefaultValue(QVariant::fromValue(xform));
+        info.isDefaultTransform = false;
+    }
 
     m_generator->generateUseNode(info);
 
-- 
2.51.2

